[ProgressBar] Empty iterable throws Exception on "maximum number of steps is not set"

freimair opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2 <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #47244 <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        |  <!-- required for new features -->

**Bug Description**
When using the progress bar with an empty input and time estimates, an error has been thrown which stopped the execution:

```
            $bar = new ProgressBar($output = $this->getOutputStream());
            $bar->setFormat("%elapsed%/%estimated%");

            $input = [];
            foreach ($bar->iterate($input) as $item)
                var_dump($item);
```

I dug around in the code and found a semantically incorrect check for `getMaxSteps()` that evaluated a maxSteps of 0 (because of an empty input) to be interpreted like maxSteps has not been set at all und thus, triggering

```
                if (!$bar->getMaxSteps())) {
                    throw new LogicException('Unable to display the .* time if the maximum number of steps is not set.');
                }
```

**Fix Proposal**

I propose to change the check to be semantically correct as in "0 is a valid value of maxSteps, null is not".

More details and a bit of discussion in #47244.

<!--
Replace this notice by a short README for your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the latest branch.
 - For new features, provide some code snippets to help understand usage.  
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

freimair committed:
Fix: ProgressBar does not fail on emtpy Iterable

freimair committed:
Fix: divide by Zero chance in ProgressBar

freimair commented:
well, the issues seems to be there since the creation of the progress bar #10356 back in 2014.

Has been merged into the 2.5-dev branch 554b28d9754521936bce928c1453e19e1da15ee3.

Please let me know if I should PR against 2.8?

fabpot commented on src/Symfony/Component/Console/Helper/ProgressBar.php:
```diff
@@ -519,14 +519,14 @@ private static function initPlaceholderFormatters(): array
             },
             'remaining' => function (self $bar) {
-                if (!$bar->getMaxSteps()) {
+                if (!isset($bar->max)) {
```
I think this should be reverted, same for below.
Also, on line 53, I think we need to change the definition of `max` to `private int $max = 0;` as `$max` can never be null (see the constructor).

freimair replied:
I do agree of course that the check will never yield true due to the constructor initialization. Have not spotted that. My bad. However, reverting the lines you mentioned would render the PR void and the issue would still exist.

Thing is, with `$max` always being an integer, there is no semantically correct way to differentiate between a progressbar with a defined maximum and a progressbar without, ie. a progressbar and a throbber, respectively.

Hence, maybe the way to go would be the other way around - keep the definition as is (`?int $max = null`) and introduce checks for `null` where needed?

Will dig into that soon. Thanks in the meanwhile.

freimair committed:
Set max to int since it cannot be null

`max` value has been prepared to contain a `null` value, however,
the constructor with its default value of `max = 0` prevents
`$max` to ever be `null`.

Hence, declare `$max` as `int` instead of `?int`.

freimair committed:
Introduce isMaxKnown variable

Up until now, it has been very hard to determin whether a progressbar
has a known maximum value or not, because the property `max` is an
integer. Up until now, `max == 0` has been treated like we do _not_
know the `max` value. However, `max == 0` _is_ a valid `max` value
when iterating over an empty but countable array for example.

We cannot simply keep `max == null` if unknown, because that would break
rendering mechanics and fixing those would require interface changes.

Hence, enter `isMaxKnown` - a property which keeps track of whether the
`max` value is known or not.

freimair committed:
Ui work

fabpot commented on src/Symfony/Component/Console/Helper/ProgressBar.php:
```diff
@@ -50,7 +50,8 @@ final class ProgressBar
     private int $startingStep = 0;
-    private ?int $max = null;
+    private int $max = 0;
+    private bool $isMaxKnown = false;
```
Do we need this new variable? I suppose `0` is not a valid max, so having max at 0 probably means that it is not known?

c0necto replied:
the whole point of this PR and the issue I found with Progessbars is that max = 0 IS a valid max value but it is treated like it is not.

we stumbled over that behavior because we need to do a lot of import/sync operations, nasty but necessary. Once in a while, one of our imports/sync iterables is empty and thus max value = 0. `foreach([] as ...)` does not throw errors in vanilla PHP. Using the progressbar with `foreach($bar->iterate([]) as ...)` does.

freimair replied:
the whole point of this PR and the issue I found with Progessbars is that max = 0 IS a valid max value but it is treated like it is not.

we stumbled over that behavior because we need to do a lot of import/sync operations, nasty but necessary. Once in a while, one of our imports/sync iterables is empty and thus max value = 0. `foreach([] as ...)` does not throw errors in vanilla PHP. Using the progressbar with `foreach($bar->iterate([]) as ...)` does.

jwilson-bm commented:
I see this PR is in milestone [6.4](https://github.com/symfony/symfony/milestone/30) - does that mean that this fix will be in the 6.4 release?

freimair committed:
Set maxKnown also in start function

GromNaN commented:
I submitted an alternative patch, using `$max = null` to indicates when the value is unknown. https://github.com/symfony/symfony/pull/52605

fabpot commented:
Closing in favor of #52605
